libraries/AP_NavEKF3/AP_NavEKF3.cpp

```diff
@@ -815,34 +822,53 @@ void NavEKF3::UpdateFilter(void)
         }
         runCoreSelection = (imuSampleTime_us - lastUnhealthyTime_us) > 1E7;
     }
-    float primaryErrorScore = core[primary].errorScore();
-    if ((primaryErrorScore > 1.0f || !core[primary].healthy()) && runCoreSelection) {
-        float lowestErrorScore = 0.67f * primaryErrorScore;
-        uint8_t newPrimaryIndex = primary; // index for new primary
-        for (uint8_t coreIndex=0; coreIndex<num_cores; coreIndex++) {
+    
+                                   
+    if(runCoreSelection) {
+        float primaryErrorScore = core[primary].errorScore();
+
+        // update the relative error scores for all active cores
+        updateCoreErrors(primaryErrorScore);
 
+        bool betterCore = false;
+        bool altCoreAvailable = false;
+        float maxCoreError = relativeCoreError[primary]; // looking for cores that have error lower than the current primary
+        uint8_t newPrimaryIndex = primary;
+
+        // loop through all available cores to find if an alternative core is available
+        for (uint8_t coreIndex=0; coreIndex<num_cores; coreIndex++) {
             if (coreIndex != primary) {
-                // an alternative core is available for selection only if healthy and if states have been updated on this time step
-                bool altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex];
-
-                // If the primary core is unhealthy and another core is available, then switch now
-                // If the primary core is still healthy,then switching is optional and will only be done if
-                // a core with a significantly lower error score can be found
-                float altErrorScore = core[coreIndex].errorScore();
-                if (altCoreAvailable && (!core[newPrimaryIndex].healthy() || altErrorScore < lowestErrorScore)) {
+                float altCoreError = relativeCoreError[coreIndex];
+
+                // an alternative core is available for selection based on 2 conditions -
+                // 1. healthy and states have been updated on this time step
+                // 2. has relative error less than primary core error
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
```

Comment:
I think we should consider dropping the statePredictEnabled check. That check made sense when we were making a decision based on a single sample, but makes less sense with the accumulated error. Paul, what do you think?

Reply:
In practice it should make little difference now given real world noise and sensor errors, but in theory if there is a sensor error transient that effects both lanes, removing it will disadvantage the leading lane. It's an easy change to test so worth a try.